### ISSUE TITLE
Refactoring a function does not fully support Rust

### ISSUE THREAD

**Issue by blueglyph** [intellij-rust/intellij-rust#8899]
## Environment

Rust plugin version 0.4.171.4656-221
Toolchain 1.61.0
IntelliJ IDEA 2022.1.1 (Ultimate Edition)
Build #IU-221.5591.52, built on May 10, 2022
Windows 10 x64

## Problem description
1) Refactoring a function's signature, in struct, trait or other, shows a 'Default Value' field. This field is ignored when refactoring, because Rust does not support default parameters (very unfortunately).

2) It is not possible to add `&self` in a trait function parameter list.

3) It is not possible to remove `&self` from a trait function parameter list.

## Steps to reproduce

- Type the definition of a new trait (`trait Test { }` will do)
- add a function with no parameter (`fn test() { }` will do)
- put the cursor in the function's signature and press F6 or select `Refactor / Change signature...`
=> there is a "Default Value" field in the parameter table
- add a `&self` parameter
=> the tool refuses to refactor if no type is given
- add a parameter with a default value, for example p0, usize, 5
- Refactor
=> the default value is ignored, there is no warning
- add the `&self` parameter manually
- put the cursor in the function's signature and press F6 or select `Refactor / Change signature...`
=> there is no way to remove `&self`

**Comment by afetisov** [comment]
You have a misunderstanding about that refactoring dialogue.

* There are no parameter default values in Rust, but if you add a parameter to a function, then all its uses would need to be updated with another argument. That's what default parameter value in the dialogue does: it provides a (supposedly valid) default value at all use sites, so that the code still compiles. Supposedly you have some valid default value to provide, otherwise the refactoring cannot be performed automatically in a non-breaking way. This is how it behaves:
```rust
fn test() {}

fn bar (){
    test();
}
```

Add parameter `p0` with type `u32` and default value `2`:

```rust
fn test(p0: u32) {}

fn bar (){
    test(2);
}
```

> add a &self parameter
> => the tool refuses to refactor if no type is given

* The `&self` parameter in methods is always implicitly typed as `&Self`. All function parameters are always explicitly typed in Rust. For that reason you cannot add a parameter without providing its type. However, the refactoring you are trying to do is invalid, since `&self` parameter can only exist on trait methods and inherent methods. If `test` is an free function, adding `&self ` would require turning it into an inherent method, which would be a very different refactoring. Changing a trait method to accept a `&self` parameter would be possible at the definition level, but the refactoring should also update all use sites, and that's impossible to do automatically. A static trait method use would look like `Foo:test()`, where `Foo` is a type implementing `Trait`. If you add `&self`, then the use would have to look like `Foo::test(&foo)`, where `foo: Foo` is some instance. There is no way to add such instance automatically at all use sites.

* Removing `&self` arguably could work, but given the above arguably it would be a weird refactoring to support. Imho converting between static and dynamic methods should be a separate refactoring, if it's even possible to do.

Overall, my opinion is "works as expected". It would be nice to add Rust-specific help link to the refactoring dialogue, if it's possible.

**Comment by blueglyph** [comment]
Now I see about the default value, thanks! I thought it was simply using the same template and that it was still WIP, my mistake.

About adding &self: the example I describe is not a stand-alone function but a trait method so it's valid.

When I try to add `self` with type `&Self` or `Self`, for example, it accepts the refactoring and shows the correct signature but the result is a method with an empty signature, it has removed the previously existing parameters instead of adding the new one, without as much as a warning. It only impacts the definition, not where the method is called.

So `fn test(p0: usize)` is then shown as `fn test(self: &Self, p0: usize)` in the dialog, and after refactoring it has become `fn test()`.

Maybe I'm missing something?

Note: `&Self` or `Self` is rarely specified, methods usually get `self` or `&self`. Since it doesn't work for me, I'm not sure whether it is supposed to add the type or not.

But granted, the refactoring has little use since the user has to manually add the instance everywhere. So maybe it's best to remove it than to fix it if it's wrong. And removing `self` would be possible but it is probably not used very often if at all, indeed.

**Comment by afetisov** [comment]
> it accepts the refactoring and shows the correct signature but the result after clicking on 'Refactor' is a method with an empty signature,

Yep, that's a bug. I'll file a more specific issue for that.

> But granted, the refactoring has little use since the user has to manually add the instance everywhere. So maybe it's best to remove it than to fix it if it's wrong.

The classic definition of refactoring is a small-step transformation between valid code. While that's not always possible or feasible, and there may be value in partially-correct refactorings, I don't think it makes sense to add something as directly broken as a refactoring to add `&self` parameter and make all calls invalid. Adding `&self` in the function definition is easy, the point of refactoring is to deal reasonably with use sites.

Imho the refactoring dialogue should explicitly reject parameters with `self` identifier as invalid. Currently there is no proper error for that case.

**Comment by afetisov** [comment]
> Yep, that's a bug. I'll file a more specific issue for that.

Fixed in #9150 

I suppose this issue can now be closed?

**Comment by blueglyph** [comment]
Yes, I'll close it. Thanks!
